File: swift/grade-school/Sources/GradeSchool/GradeSchool.swift
typealias Student = String
typealias Grade   = Int
typealias Roster  = [Grade: [Student]]

class GradeSchool {
    private(set) var roster: Roster
    
    init() { self.roster = [:] }

    func addStudent(_ name: Student, grade g: Grade) -> Void {
        var grade = self.studentsInGrade(g)
        grade.append(name)
        self.roster[g] = grade
    }

    func studentsInGrade(_ g: Grade) -> [Student] {
        return self.roster[g] ?? []
    }

    var sortedRoster: Roster {
        get {
            var sorted: Roster = [:]
            for (grade, students) in self.roster {
                sorted[grade] = students.sorted()
            }
            return sorted
        }
    }
}

File: dart/grade-school/lib/grade_school.dart
typedef Student = (String, int);

class GradeSchool {
  List<Student> directory = [];

  List<String> roster() => _names(directory);

  List<String> _names(Iterable<Student> students) =>
      students.map((s) => s.$1).toList();

  List<bool> add(List<Student> students) => students.map(_addStudent).toList();

  bool _addStudent(Student student) {
    if (directory.any((s) => s.$1 == student.$1)) {
      return false;
    }

    directory.add(student);
    directory.sort((a, b) {
      var cmp = a.$2.compareTo(b.$2);
      return cmp != 0 ? cmp : a.$1.compareTo(b.$1);
    });
    return true;
  }

  List<String> grade(int grade) =>
      _names(directory.where((s) => s.$2 == grade));
}
